File: MusicAlbums/MusicAlbums/ViewController/ViewController.m
//
//  ViewController.m
//  MoviesObjC
//
//  Created by iAskedYou2nd on 8/17/20.
//  Copyright © 2020 iAskedYou2nd. All rights reserved.
//

#import "ViewController.h"
#import "PokeCell.h"
#import "ResponseObj.h"
#import "NetworkManager.h"
#import "Pokemon.h"
#import "MusicAlbums-Swift.h"

@interface ViewController ()

@property (nonatomic, weak) UITableView* tableView;
@property (nonatomic, strong) PokemonViewModel* vm;


@end


@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setUp];
        
    //[self.vm fetchPokes];
    [self setUpViewModel];
    [self.vm fetchPokes];
}

- (void)setUp {
    UITableView* table = [[UITableView alloc] initWithFrame:CGRectZero];
    [table setTranslatesAutoresizingMaskIntoConstraints:false];
    [table setDataSource:self];
    // Set Prefetch
    //[self.tableView setPrefetchDataSource:self];
    // Register Cell
    [table registerClass:[PokeCell self] forCellReuseIdentifier:@"CellId"];
    
    self.tableView = table;
    [self.view addSubview:table];
    
    // Constraints
    [[table.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:8] setActive:true];
    [[table.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor constant:8] setActive:true];
    [[table.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor constant:-8] setActive:true];
    [[table.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-8] setActive:true];
    
    [table setBackgroundColor:UIColor.systemTealColor];
}

-(void)setUpViewModel {
    self.vm = [[PokemonViewModel alloc] initWithService:NetworkManager.sharedInstance];
    
    [self.vm bindWithCompletion:^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];
        });
    }];
}


// MARK: - DataSource Functions

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.vm.count;
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    PokeCell* cell = (PokeCell*)[tableView dequeueReusableCellWithIdentifier:@"CellId" forIndexPath:indexPath];
    if (cell == nil) {
        return [[UITableViewCell alloc] init];
    }
    
    // Movie set up for cells
    
    //[self.vm fetchSpritesWithIndex:indexPath.row];
    [self.vm fetchSpritesWithIndex:indexPath.row completion:^(UIImage * _Nullable image) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [cell.pokeName setText:[self.vm titleWithIndex:indexPath.row]];
            [cell.img setImage:image];
            
        });
    }];
    
    return cell;
}

// MARK: - Prefetch DataSource

//-(void)tableView:(UITableView *)tableView prefetchRowsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths {
//    NSIndexPath* lastIndexPath = [NSIndexPath indexPathForRow:self.vm.count - 1 inSection:0];
//    if ([indexPaths containsObject:lastIndexPath]) {
//        [self.vm fetchPokes];
//    }
//}

@end

File: JSONDragon/JSONDragon/JSONDragon/ViewController/ViewController.swift
import UIKit

class ViewController:  UIViewController, UITableViewDelegate, UITableViewDataSource {

    var pokemonType: PokemonType?
    var dragonsList: [String] = []
    var tableView: UITableView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.pokemonType = NetworkManager.shared.getPokemonType()
        guard let pokemonType = self.pokemonType else { return }
        let dragons = pokemonType.dragonList
               dragons.forEach{
                   self.dragonsList.append($0.pokemon.name)
                   print($0.pokemon.name)
               }
//       print(dragonsList)
//       print("done")
        self.setUp()
    }
    
    private func setUp() {
        let tableView = UITableView(frame: .zero)
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(DragonCellTableViewCell.self, forCellReuseIdentifier: DragonCellTableViewCell.reuseId)
        tableView.backgroundColor = .white
        
        self.view.addSubview(tableView)
        
        tableView.topAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.topAnchor, constant: 8).isActive = true
        tableView.leadingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.leadingAnchor, constant: 8).isActive = true
        tableView.trailingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.trailingAnchor, constant: -8).isActive = true
        tableView.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor, constant: -8).isActive = true
            
        self.tableView = tableView
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dragonsList.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: DragonCellTableViewCell.reuseId, for: indexPath) as? DragonCellTableViewCell else {
            return UITableViewCell()
        }
        
        cell.dragonName?.text = dragonsList[indexPath.row]
        
        return cell
    }

}
